src/java/org/apache/cassandra/schema/Types.java

```diff
@@ -498,16 +572,95 @@ public long serializedSize(Types t, Version version)
         }
+
+        private void serializeFieldMetadata(UserType type, List<FieldIdentifier> fieldIdentifiers, DataOutputPlus out) throws IOException
+        {
+            // Sparse serialization: only serialize non-empty metadata with their positions
+            serializeSparseMetadata(fieldIdentifiers, type::fieldComment, out);
+            serializeSparseMetadata(fieldIdentifiers, type::fieldSecurityLabel, out);
+        }
+
+        private void serializeSparseMetadata(List<FieldIdentifier> fieldIdentifiers,
+                                             Function<FieldIdentifier, String> metadataGetter,
+                                             DataOutputPlus out) throws IOException
+        {
+            // Collect only non-empty indexes
+            List<Integer> nonEmptyIndexes = new ArrayList<>();
+            for (int i = 0; i < fieldIdentifiers.size(); i++)
+            {
+                String value = metadataGetter.apply(fieldIdentifiers.get(i));
+                if (!value.isEmpty())
+                    nonEmptyIndexes.add(i);
+            }
+
+            // Write count followed by position-value pairs
+            out.writeUnsignedVInt32(nonEmptyIndexes.size());
+            for (int index : nonEmptyIndexes)
+            {
+                out.writeUnsignedVInt32(index);
+                out.writeUTF(metadataGetter.apply(fieldIdentifiers.get(index)));
+            }
+        }
+
+        private void deserializeFieldMetadata(DataInputPlus in, List<String> fieldComments, List<String> fieldSecurityLabels) throws IOException
+        {
+            deserializeSparseMetadata(in, fieldComments);
+            deserializeSparseMetadata(in, fieldSecurityLabels);
+        }
+
+        private void deserializeSparseMetadata(DataInputPlus in, List<String> target) throws IOException
+        {
+            int count = in.readUnsignedVInt32();
+            for (int i = 0; i < count; i++)
+            {
+                int position = in.readUnsignedVInt32();
+                String value = in.readUTF();
+                target.set(position, value);
+            }
+        }
+
+        private long fieldMetadataSize(UserType type, List<FieldIdentifier> fieldIdentifiers)
+        {
+            return sparseMetadataSize(fieldIdentifiers, type::fieldComment)
+                 + sparseMetadataSize(fieldIdentifiers, type::fieldSecurityLabel);
+        }
+
+        private long sparseMetadataSize(List<FieldIdentifier> fieldIdentifiers,
+                                       java.util.function.Function<FieldIdentifier, String> metadataGetter)
+        {
+            int nonEmptyCount = 0;
+            long dataSize = 0;
+
+            for (FieldIdentifier fieldId : fieldIdentifiers)
+            {
+                String value = metadataGetter.apply(fieldId);
+                if (!value.isEmpty())
+                {
+                    nonEmptyCount++;
+                    dataSize += sizeofUnsignedVInt(0);     // position (int)
```

Comment:
You cannot get the size of 0 for all items. `sizeofUnsignedVInt` can return different values based on the input. 
Please use the actual index value, instead of 0.